<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="img/favicon.ico">
  <title>Usage guideline - CusProSe documentation</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="css/theme.css" />
  <link rel="stylesheet" href="css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  <link href="css/extra.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Usage guideline";
    var mkdocs_page_input_path = "psd_usage.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="js/jquery-2.1.1.min.js" defer></script>
  <script src="js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="." class="icon icon-home"> CusProSe documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="index.html">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Iterhmmbuild</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="ihb_introduction.html">Overall procedure</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="ihb_usage.html">Usage guideline</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Prosecda</span></p>
                <ul class="current">
                    <li class="toctree-l1"><a class="reference internal" href="psd_introduction.html">Overall procedure</a>
                    </li>
                    <li class="toctree-l1 current"><a class="reference internal current" href="psd_usage.html">Usage guideline</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#quick-start-example">Quick start example</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#creating-rules">Creating rules</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#edit-rules-from-gui">Edit rules from GUI</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#output-of-prosecda">Output of ProSeCDA</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#overall-architecture">Overall architecture</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#content-of-the-results-directory">Content of the results/ directory</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#illustration-of-the-output-files">Illustration of the output files</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#the-all_familynamepdf-file">The all_familyName.pdf file</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#all_familynamepdf-vs-mgg_xxxpdf">all_familyName.pdf VS MGG_xxx.pdf</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#xml-file">XML file</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href=".">CusProSe documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".">Docs</a> &raquo;</li>
    
      
        
          <li>Prosecda &raquo;</li>
        
      
    
    <li>Usage guideline</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="prosecda-usage-guideline">ProSeCDA usage guideline<a class="headerlink" href="#prosecda-usage-guideline" title="Permanent link">#</a></h1>
<div class="admonition tip">
    <p class="first admonition-title">
        Tip
    </p>
    <p>
    To guide the user, input data files can be found in <code>cusProSe/prosecda/datas/</code>:
    <pre style="line-height: 15px;"><b>cusProSe/prosecda/datas/</b>
      ├── database.hmm
      ├── mgg_70-15_8.fasta
      └── rules.yaml
    </pre>
    <ul>
    Those are the three mandatory files required as inputs to run prosecda:
    <li><code>rules.yaml</code> is a YAML file in which proteins of interest are described based on the presence/absence of specific domains</li>
    <li><code>database.hmm</code> is a customized HMM profile database made of 24 HMM protein domain models</li>
    <li><code>mgg_70-15_8.fasta</code> is the <i>magnaporthe orizae</i> proteome that will be scanned against proteins of interest defined in the YAML rules file</li>
    </ul>
    </p>
</div>

<h2 id="quick-start-example">Quick start example<a class="headerlink" href="#quick-start-example" title="Permanent link">#</a></h2>
<div>
The command below will search in the <i>magnaporthe orizae</i> proteome all proteins defined in <code>rules.yaml</code> (the HMM profile database <code>database.hmm</code> is used to find domains of interest matching proteins in the <i>magnaporthe orizae</i> proteome):

<pre><code class="language-bash">prosecda -proteome mgg_70-15_8.fasta -hmmdb database.hmm -rules rules.yaml
</code></pre>

</div>

<p><br>
Help about the usage of ProSeCDA and its parameters can be shown with the following command: <code>prosecda -h</code>:</p>
<pre class="parameters">usage: prosecda [-h] -proteome [PROTEOME] -hmmdb [HMMDB] -rules [RULES] [-out [OUT]] 
                [-cov COV] [-cevalue CEVALUE] [-ievalue IEVALUE] [-score SCORE] [-acc ACC]
                [--nopdf]

Search proteins matching rules.

optional arguments:
  -h, --help            show this help message and exit
  -proteome [PROTEOME]  Proteome file (.fasta)
  -hmmdb [HMMDB]        HMM profile database
  -rules [RULES]        Rules&apos;file (.yaml)
  -out [OUT]            Output directory
  -cov COV              Minimum ratio between the length of the HMM profile stretch that 
                        matches a sequence and the overall length of the HMM profile (0.0)
  -cevalue CEVALUE      HMMER conditional e-value cutoff (0.01)
  -ievalue IEVALUE      HMMER independant e-value cutoff (0.01)
  -score SCORE          HMMER score cutoff (3.0)
  -acc ACC              HMMER mean probability of the alignment accuracy between each residues
                        of the target and the corresponding hmm state (0.6)
  --nopdf               Deactivate the generation of the pdf results (False)
</pre>

<h2 id="creating-rules">Creating rules<a class="headerlink" href="#creating-rules" title="Permanent link">#</a></h2>
<p>The proteins the user is interested in must be described in a file containing simple rules defining for each protein of interest which domains are mandatory and/or which one are forbidden. An E-value (corresponding to the independent domain E-value from hmmsearch) can also be specified for each mandatory domains. In that case, a mandatory domain must match a sequence with an E-value at least below the one specified. If no E-value is defined for mandatory domains, an E-value of 0.01 is used by default.</p>
<h4 id="edit-rules-from-gui">Edit rules from GUI<a class="headerlink" href="#edit-rules-from-gui" title="Permanent link">#</a></h4>
<p>The file in which rules are edited must respect a specific syntax in a YAML format. To make it simpler, a Graphical User Interface (GUI) has been implemented. You can access it through the command <code>create_rules</code></p>
<figure class="fig-prosecda">
    <img src="./img/prosecda_rules.png"
      alt="Graphical User Interface to edit rules.">
    <figcaption>
<b>Figure 3: Graphical User Interface to edit rules</b> 
    </figcaption>
</figure>

<p>As shown in the above image, a rule can be defined by five different inputs:</p>
<ul>
<li><code>Name</code> (required): category/family name the proteins that'll match the rule will be assigned to</li>
<li><code>Comment</code> (optional): can be used to describe the rule</li>
<li><code>Mandatory</code> (required): domain name the protein must contain</li>
<li><code>E-value</code> (optional): minimal E-value the associated domain must match a sequence (0.01 if not specified)</li>
<li><code>Forbidden</code> (optional): domain name the protein must not contain</li>
</ul>
<p>Once you have added all your mandatory/forbidden domains describing the domain architecture your proteins of interest must match, click on <code>Add rule to the list</code> to validate the rule. You'll see it listed in the <code>Rule list</code> panel. Repeat the procedure to create as many rules as you want. </p>
<p>Once all your desired rules are defined, click on <code>Save rules</code> to save them in a YAML file. This YAML file will look like the one described below:</p>
<pre><code class="language-yaml">PKS:
 COMMENT: Polyketide Synthase
 CONDITION:
  mandatory:
  - KS
  - AT
  - PP-binding, 0.0001
  forbidden:
  - C

PKS-like:
 COMMENT: Polyketide Synthase like
 CONDITION:
  mandatory:
  - KS
  - AT
  forbidden:

PKS_type3.0:
 COMMENT: Polyketide Syntase type III
 CONDITION:
  mandatory:
  - CHS_like
  forbidden:

</code></pre>
<h2 id="output-of-prosecda">Output of ProSeCDA<a class="headerlink" href="#output-of-prosecda" title="Permanent link">#</a></h2>
<h4 id="overall-architecture">Overall architecture<a class="headerlink" href="#overall-architecture" title="Permanent link">#</a></h4>
<p>The output directory generated from the command run in the <a href="#quick-start-example">Quick start example</a> will be as described below:</p>
<pre><b>prosecda_year-month-day_hour-min-sec/</b>
├── <b>css/</b>
├── <b>images/</b>
├── index.html
├── info.log
├── <b>js/</b>
├── mgg_70-15_8.domtblout
└── <b>results/</b>
 </pre>

<p>with:</p>
<ul class="myul2">
  <li><code>index.html</code>: an interactive web page allowing to visualize the results (described <a href="#a">here</a>)</li>
  <li><code>css/</code>, <code>js/</code> and <code>images/</code>: folders containing files read by index.html</li>
  <li><code>info.log</code>: summary log of the computation run</li>
  <li><code>mgg_70-15_8.domtblout</code>: output file of hmmsearch</li>
  <li><code>results/</code>: directory containing different output files for rules matching proteins</li>
</ul>

<h4 id="content-of-the-results-directory">Content of the results/ directory<a class="headerlink" href="#content-of-the-results-directory" title="Permanent link">#</a></h4>
<p>The <code>results/</code> directory contains a list of subdirectories, each one corresponding to a user-defined protein family for which match has been found (<code>DMATS/</code>, <code>Ent_kaurene_synthase/</code> ...):</p>
<pre><b>results/</b>
├── <b>DMATS/</b>
│   ├── all_DMATS.pdf
│   ├── MGG_06540T0.fa
│   ├── MGG_06540T0.pdf
│   ├── MGG_06540T0.xml
│   ├── MGG_10953T0.fa
│   ├── MGG_10953T0.pdf
│   ├── MGG_10953T0.xml
│   ├── MGG_12480T0.fa
│   ├── MGG_12480T0.pdf
│   └── MGG_12480T0.xml
├── <b>Ent_kaurene_synthase/</b>
│   ├── all_Ent_kaurene_synthase.pdf
│   ├── MGG_01949T0.fa
│   ├── MGG_01949T0.pdf
│   ├── MGG_01949T0.xml
|   ...
... 
</pre>

<p>Inside each protein family folder, three files with the related protein ID as the basename are generated for each protein matching the user-defined family rule:</p>
<ul class="myul2">
  <li><code>MGG_xxx.fa</code>: the protein sequence in fasta format</li>
  <li><code>MGG_xxx.xml</code>: details in XML format about the proteins and its matching domains</li>
  <li><code>MGG_xxx.pdf</code>: a graphical representation of all of the domains that matched the protein</li>
</ul>

<p>The file whose basename follows the syntax <code>all_familyName.pdf</code> is a graphical representation of the most-likely domains architecture for all the proteins that have been found to match the user-defined family rule. Please note that for a given protein the graphical representation in this file will be different from the one found in <code>MGG_xxx.pdf</code>, the latter representing all the domains that have been found to match the protein. Some of those domains might not be visible in the most-likely domain architecture. This will be the case for a domain <i>i</i> that overlaps with a domain <i>j</i> and if any architecture in which domain <i>i</i> is present is not the most-likely architecture, that is the one with the highest score (see <a href="./psd_introduction.html#resolving-overlapping-domains">here</a> for details about how the score is assigned).</p>
<h2 id="illustration-of-the-output-files">Illustration of the output files<a class="headerlink" href="#illustration-of-the-output-files" title="Permanent link">#</a></h2>
<h4 id="the-all_familynamepdf-file">The all_familyName.pdf file<a class="headerlink" href="#the-all_familynamepdf-file" title="Permanent link">#</a></h4>
<p>The file whose basename follows the syntax <code>all_familyName.pdf</code> is a multipage pdf file containing graphical representations of the most-likely domain architecture of all the proteins that have been found to match the user-defined family rule.</p>
<p>An example of the first page of <code>all_PKS.pdf</code> is shown below:</p>
<figure >
    <img src="./img/pks_mla_examples.png"
      alt="IterHMMBuild pipeline overview.">
</figure>

<p>The first line of the figure represents all domains composing the most-likely domain architecture found for the protein MGG_00233T0. The next lines represent details information about each domain, namely its amino acid positions, name, independant e-value (i-evalue) and its score.</p>
<h4 id="all_familynamepdf-vs-mgg_xxxpdf">all_familyName.pdf VS MGG_xxx.pdf<a class="headerlink" href="#all_familynamepdf-vs-mgg_xxxpdf" title="Permanent link">#</a></h4>
<p>At the difference of the <code>all_familyName.pdf</code>, the graphical representation found for a given protein in the file <code>MGG_xxx.pdf</code> does not show any domain architecture but instead represents all the domains that matched the protein sequence during the <a href="psd_introduction.html#annotation">annotation step</a>. It can be helpful to visualize overlapping domains that have not been retained to be a part of the most-likely domain architecture. It thus provides the user a way to check the validity for a protein to really belong to the assigned protein family.</p>
<p>To illustrate that point, let's compare both graphical representations for the protein MGG_01949T0 that has been assigned to the Ent_kaurene_synthase family:</p>
<figure>
    <img  style="border: 1px solid" src="./img/ent_kaurene_MLA_VS_domains.png"
      alt="IterHMMBuild pipeline overview.">
    <figcaption>
    <b>Figure 4:</b> Comparison of the graphical representations of the protein MGG_1949T0 found in (<b>A</b>) <code>all_Ent_kaurene_synthase.pdf</code> and (<b>B</b>) <code>MGG_01949T0.pdf</code>.
    </figcaption>
</figure>

<p>The figure 4A shows the most-likely domain architecture assigned to the protein MGG_01949T0 (found in <code>all_Ent_kaurene_synthase.pdf</code>) which is composed of only one domain: the <i>ent_kaurene</i> domain. </p>
<p>The figure 4B shows all the domains that matched the protein sequence during the annotation step. We can see in this representation that both the <i>ent_kaurene</i> domain and the <i>diterpene</i> domain matched the protein sequence with an e-value equals to 0.0 and a really high score. Although the former has been retained in the most-likely domain architecture (based on its slightly higher score) and thus made the protein assigned as an Ent_kaurene_synthase, this choice might be questionable based on the similarly high score for the latter domain. </p>
<h3 id="xml-file">XML file<a class="headerlink" href="#xml-file" title="Permanent link">#</a></h3>
<pre><code class="language-xml">&lt;protein&gt;
  &lt;id&gt;XP_023431478.1&lt;/id&gt;
  &lt;sequence_length&gt;952&lt;/sequence_length&gt;
  &lt;class_name&gt;Ent_kaurene_synthase&lt;/class_name&gt;
  &lt;domain_architecture&gt;
    &lt;domain name=&quot;ent_kaurene&quot;&gt;
      &lt;cval&gt;0.0&lt;/cval&gt;
      &lt;ival&gt;0.0&lt;/ival&gt;
      &lt;score&gt;1074.3&lt;/score&gt;
      &lt;start&gt;10&lt;/start&gt;
      &lt;end&gt;952&lt;/end&gt;
      &lt;domain_length&gt;943&lt;/domain_length&gt;
    &lt;/domain&gt;
  &lt;/domain_architecture&gt;
&lt;/protein&gt;
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="psd_introduction.html" class="btn btn-neutral" title="Overall procedure"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="psd_introduction.html" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme.js" defer></script>
      <script src="search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
